### PR TITLE
[Merged by Bors] - feat: sub/superscript parser

### PR DESCRIPTION
**Pull request by digama0**
This is intended for use in local notations, like this:
```lean
import Mathlib.Util.Superscript

local syntax:arg term:max superscript(term) : term
local macro_rules | `($a:term $b:superscript) => `($a ^ $b)

example : 2² = 4 := rfl
example : 2¹⁶ = 65536 := rfl
example (n : Nat) : n⁽²⁻¹⁾ ⁺ ⁶ /- not done yet... -/ ⁺ ᶠᵒᵒ = n ^ (7 + foo) := rfl
example : (fun n => 2ⁿ⁺¹) 15 = 2¹⁶ := rfl
```
See also the discussion on [Zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/superscript.20parser/near/365151927).

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Util/Superscript.lean [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+
+namespace Mathlib.Tactic
+open Lean Parser PrettyPrinter
+namespace Superscript
+
+instance : Hashable Char := ⟨fun c => hash c.1⟩
+
+/-- A bidirectional character mapping. -/
+structure Mapping where
+  /-- Map from "special" (i.e. superscript) characters to "normal" characters. -/
```
Comment (fpvandoorn) [inline Mathlib/Util/Superscript.lean:36]:
```suggestion
  /-- Map from "special" (e.g. superscript) characters to "normal" characters. -/
```

File: Mathlib/Util/Superscript.lean [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+  toNormal : HashMap Char Char := {}
+  /-- Map from "normal" text to "special" (i.e. superscript) characters. -/
```
Comment (fpvandoorn) [inline Mathlib/Util/Superscript.lean:38]:
```suggestion
  /-- Map from "normal" text to "special" (e.g. superscript) characters. -/
```

**Review by fpvandoorn — COMMENTED** [review]
LGTM

There are still some FIXME and deleteme in the code. Do you want to merge before fixing them?

bors d+

**Comment by digama0** [comment]
* The FIXME is about having a formatter for the parser. Currently if you use the pretty printer it will print `2³` as `2 3`, ignoring the superscriptness. I will leave it for a future PR.

* The deleteme section is about leanprover/lean4#2269, it can just be removed once that PR lands and doesn't really block this one.

bors r+

**Pull request closed without merging**